lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/EvaluatorManager.cs

```diff
@@ -0,0 +1,317 @@
+        ///   The number of AllocatedEvaluators has reached the total expected Evaluators
+        /// </summary>
+        /// <param name="evaluator"></param>
+        internal void AddAllocatedEvaluator(IAllocatedEvaluator evaluator)
+        {
+            if (IsAllocatedEvaluator(evaluator.Id))
+            {
+                string msg = string.Format("The allocated evaluator {0} already exists.", evaluator.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (IsEvaluatorForMaster(evaluator))
+            {
+                SetMasterEvaluatorId(evaluator.Id);
+            }
+
+            if (_masterEvaluatorId == null && NumberOfAllocatedEvaluators + 1 == _totalExpectedEvaluators)
+            {
+                string msg = string.Format("Trying to add the last Evaluator {0} but master Evaluator is not added yet.", evaluator.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (NumberOfAllocatedEvaluators == _totalExpectedEvaluators)
+            {
+                string msg = string.Format("Trying to add an additional Evaluator {0}, but the total expected Evaluator number {1} has been reached.", evaluator.Id, _totalExpectedEvaluators);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+            _allocatedEvaluators.Add(evaluator.Id);
+        }
+
+        /// <summary>
+        /// Remove an Evaluator from allocated Evaluator collection by evaluator id.
+        /// If the given evaluator id is not in allocated Evaluator collection, throw IMRUSystemException.
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        internal void RemoveAllocatedEvaluator(string evaluatorId)
+        {
+            if (!IsAllocatedEvaluator(evaluatorId))
+            {
+                string msg = string.Format("The allocated evaluator to be removed {0} does not exist.", evaluatorId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+            _allocatedEvaluators.Remove(evaluatorId);
+        }
+
+        /// <summary>
+        /// Returns number of allocated Evaluators
+        /// </summary>
+        internal int NumberOfAllocatedEvaluators
+        {
+            get { return _allocatedEvaluators.Count; }
+        }
+
+        /// <summary>
+        /// Checks if the Evaluator with the specified id has been allocated
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        /// <returns></returns>
+        internal bool IsAllocatedEvaluator(string evaluatorId)
+        {
+            return _allocatedEvaluators.Contains(evaluatorId);
+        }
+
+        /// <summary>
+        /// Checks if all the expected Evaluators are allocated.
+        /// </summary>
+        internal bool AreAllEvaluatorsAllocated
```

Comment:
One of the principle of using property vs. method is "They should not have dependencies on each other.". Looks like it is safer to use method for those that have computing which has dependency on other properties. I will change those unless it only returns a value.